pdd/trainer.py

```diff
@@ -0,0 +1,161 @@
+    X = []
+    y = []
+    with torch.no_grad():
+        for batch_idx, (inputs, targets) in enumerate(loader):
+            inputs, targets = inputs.to(device), targets.to(device)
+            outputs = model(inputs).detach().cpu().numpy()
+            targets = targets.detach().cpu().numpy()
+            if (outputs.shape[0] == batchsize):
+                X.append(outputs)
+                y.append(targets)
+    return np.vstack(X), np.hstack(y)
+
+
+def save_model(model, optimizer):
+    torch.save(model, 'triplet_model.pt')
+    torch.save(model.state_dict(), 'triplet_weight.pt')
+    torch.save(optimizer.state_dict, 'optim.pt')
+
+
+class TripletTrainer(object):
+    def __init__(self,
+                 model,
+                 optimizer,
+                 tri_train_load,
+                 epochs,
+                 tri_test_load,
+                 batch_size,
+                 KNN_train_data_load,
+                 KNN_test_data_load,
+                 scheduler,
+                 nameofplotclasses,
+                 num_classes,
+                 miner,
+                 loss_history
+                 ):
+        self.device = 'cuda' if torch.cuda.is_available() else 'cpu'
+        self.model = model.to(self.device)
+        self.optimizer = optimizer
+        # self.loss = loss
+        self.epochs = epochs
+        self.tri_train_load = tri_train_load
+        self.tri_test_load = tri_test_load
+        self.batch_size = batch_size
+        self.train_loader = KNN_train_data_load
+        self.test_loader = KNN_test_data_load
+#         self.scheduler=scheduler
+        self.nameofplotClasses = nameofplotclasses
+        self.num_classes = num_classes
+        self.miner = miner
+        self.loss_history = loss_history
+
+    def train(self):
+        for e in tqdm(range(self.epochs), desc='Epoch'):
+
+            test_em, test_labels = forward_inputs_into_model(self.test_loader, self.model,
+                                                             self.device, self.batch_size)
+            train_em, train_labels = forward_inputs_into_model(self.train_loader, self.model,
+                                                               self.device, self.batch_size)
+            knn_acc(
+                test_em,
+                test_labels,
+                train_em,
+                train_labels,
+                COUNT_NEIGHBOR_EXP_1,
+                METRIC_KNN)
+            knn_acc(
+                test_em,
+                test_labels,
+                train_em,
+                train_labels,
+                COUNT_NEIGHBOR_EXP_2,
+                METRIC_KNN)
+
+            plot(
+                train_em,
+                train_labels,
+                self.nameofplotClasses,
+                'train_embeddings',
+                COLORS)
+            plot(
+                test_em,
+                test_labels,
+                self.nameofplotClasses,
+                'test_embeddings',
+                COLORS)
```

Comment:
duplication of code. Iterate through ['train', 'test'] or choose another strategy. The same for above lines